### PR TITLE
Bug fixes and refactoring of cob_omni_drive_controller

### PR DESCRIPTION
**Pull request by mathias-luedtke**
This PR contains and supersedes #23 
- bug fixes
- improved rate limiting ( set max rate to zero to turn it off)
- separated UndercarriageGeom and UndercarriageCtrl
- Parameter parsing with URDF support
- configuration examples
- improved WheelController

Please note: cob_undercarraige_ctrl_new currently conflicts with this update due to separation of geometry and control params, but can be migrated easily.

### CODE REVIEW

File: cob_omni_drive_controller/src/param_parser.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        return false;
+    }
+
+    for(int i = 0; i < wheel_list.size(); ++i){
+
+        W params;
+        XmlRpc::XmlRpcValue & wheel = wheel_list[i];
+        MergedXmlRpcStruct merged(wheel, defaults);
+
+        if(parseWheel(params, wheel, merged, has_model?&model:0)){
```
Comment (fmessmer) [inline cob_omni_drive_controller/src/param_parser.cpp]: must be !parseWheel(params, wheel, merged, has_model?&model:0)

File: cob_omni_drive_controller/src/param_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        && read(params.dDamp, "damp", steer)
+        && read(params.dVirtM, "virt_mass", steer)
+        && read(params.dDPhiMax, "d_phi_max", steer)
+        && read(params.dDDPhiMax, "dd_phi_max", steer);
+}
+
+bool parseWheelGeom(UndercarriageGeom::WheelGeom & geom, XmlRpc::XmlRpcValue &wheel, MergedXmlRpcStruct &merged, urdf::Model* model){
+
+    try_read(geom.steer_name, "steer_name", wheel);
+    try_read(geom.drive_name, "drive_name", wheel);
```
Comment (fmessmer) [inline cob_omni_drive_controller/src/param_parser.cpp]: discrepancy between 'steer_name' here and 'steer' in the example yaml's

File: cob_omni_drive_controller/src/odom_plugin.cpp [hunk 3]
```diff
@@ -44,7 +44,6 @@ class OdometryController: public GeomController<hardware_interface::JointStateIn
         return true;
```
Comment (fmessmer) [inline cob_omni_drive_controller/src/odom_plugin.cpp:51]: forgot to advertise odometry topic, add:

topic_pub_odometry_ = controller_nh.advertise<nav_msgs::Odometry>("odom", 1);

Comment (fmessmer) [inline cob_omni_drive_controller/src/odom_plugin.cpp:51]: in `createTimer` use `1/publish_rate`

File: cob_omni_drive_controller/src/param_parser.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+template<typename T> bool try_read(T& val, const std::string &name, XmlRpc::XmlRpcValue &wheel, const T& def=T()){
+    if(wheel.hasMember(name)){
+        val = (T)wheel[name];
+        return true;
+    }else{
+        val = def;
+        return false;
+    }
+}
+
```
Comment (fmessmer) [inline cob_omni_drive_controller/src/param_parser.cpp:52]: use 

```
template<typename T> bool read_optional(T& val, const std::string &name, XmlRpc::XmlRpcValue &wheel, const T& def=T()){
    if(wheel.hasMember(name)){
        val = (T)wheel[name];
        return true;
    }
}
```

for x_pos, y_pos and wheel_radius...!

File: cob_omni_drive_controller/src/UndercarriageCtrlGeom.cpp [hunk 5]
```diff
@@ -194,145 +213,55 @@ void UndercarriageCtrlGeom::WheelData::calcControlStep(WheelState &command, doub
...
 
 }
-
-void UndercarriageCtrlGeom::reset()
-{
-    for(size_t i = 0; i < wheels_.size(); ++i){
-        wheels_[i].reset();
-    }
-
-}
```
Comment (fmessmer) [inline cob_omni_drive_controller/src/UndercarriageCtrlGeom.cpp:290]: do not remove reset()

**Comment by mathias-luedtke** [comment]
URDF parsing works now.
We were able to start the controllers in simulation, but the wheels did not move properly (moving steers but not drives).

@ipa-fxm: I have decided to name the joint parameters "steer" and "drive", as in the examples

**Comment by fmessmer** [comment]
Thanks @ipa-mdl , looks good now!

Tested in simulation with cob3-6 and cob4-2...looks like it's doing something plausible when sending twists...:wink: 

However, final confirmation should come from a Nav-Team member - after severly testing it with cob_navigation/nav2.0...

Will `/base/wheel_controller/command` be the final name of the topic for sending twists?
If we go on using this plugin and this name, we would need to adjust teleop-tools (e.g. teleop_keyboard, velocity_smoother, ...)

I can propose PR for cob_robots where this plugins are used in gazebo simulation!
Should I do this for all robots? cob3-6, cob3-9, cob4-1, cob4-2,..., raw's?

**Comment by mathias-luedtke** [comment]
According to our meeting regarding the namespaces, the topic will be /base/twist_controller/command[_safe], at least this was written on the whiteboard ;)
cob4-6 was already moving running the new controller, but we need further tests.

**Comment by floweisshardt** [comment]
visual review: OK

**Pull request merged**